app/src/main/java/org/schabi/newpipe/local/bookmark/BookmarkFragment.java

```diff
@@ -247,12 +314,242 @@ protected void resetFragment() {
+                isDisplayIndexModified = true;
+            }
+
+            // Updating display index in the item does not affect the value inserts into
+            // database, which will be recalculated during the database update. Updating
+            // display index in the item here is to determine whether it is recently modified.
+            // Save the index read from the database.
+            if (item instanceof PlaylistMetadataEntry) {
+
+                displayIndexInDatabase.put(new Pair<>(((PlaylistMetadataEntry) item).uid,
+                        LocalItem.LocalItemType.PLAYLIST_LOCAL_ITEM), item.getDisplayIndex());
+                ((PlaylistMetadataEntry) item).displayIndex = i;
+
+            } else if (item instanceof PlaylistRemoteEntity) {
+
+                displayIndexInDatabase.put(new Pair<>(((PlaylistRemoteEntity) item).getUid(),
+                                LocalItem.LocalItemType.PLAYLIST_REMOTE_ITEM),
+                        item.getDisplayIndex());
+                ((PlaylistRemoteEntity) item).setDisplayIndex(i);
+
+            }
+        }
+
+        if (debounceSaver != null && isDisplayIndexModified) {
+            debounceSaver.saveChanges();
+        }
+    }
+
+    @Override
+    public void saveImmediate() {
+        if (itemListAdapter == null) {
+            return;
+        }
+
+        // List must be loaded and modified in order to save
+        if (isLoadingComplete == null || debounceSaver == null
+                || !isLoadingComplete.get() || !debounceSaver.getIsModified()) {
+            Log.w(TAG, "Attempting to save playlists in bookmark when bookmark "
+                    + "is not loaded or playlists not modified");
+            return;
+        }
+
+        final List<LocalItem> items = itemListAdapter.getItemsList();
+        final List<PlaylistMetadataEntry> localItemsUpdate = new ArrayList<>();
+        final List<Long> localItemsDeleteUid = new ArrayList<>();
+        final List<PlaylistRemoteEntity> remoteItemsUpdate = new ArrayList<>();
+        final List<Long> remoteItemsDeleteUid = new ArrayList<>();
+
+        // Calculate display index
+        for (int i = 0; i < items.size(); i++) {
+            final LocalItem item = items.get(i);
+
+            if (item instanceof PlaylistMetadataEntry) {
+                ((PlaylistMetadataEntry) item).displayIndex = i;
+
+                final Long uid = ((PlaylistMetadataEntry) item).uid;
+                final Pair<Long, LocalItem.LocalItemType> key = new Pair<>(uid,
+                        LocalItem.LocalItemType.PLAYLIST_LOCAL_ITEM);
+                final Long databaseIndex = displayIndexInDatabase.remove(key);
+
+                // The database index should not be null because inserting new item into database
+                // is not handled here. NullPointerException has occurred once, but I can't
+                // reproduce it. Enhance robustness here.
+                if (databaseIndex != null && databaseIndex != i) {
```

Comment:
I don't understand why you need to store the original database index and all of the other data. I would do it like this:
- when initially receiving the items, just set them in the list
- keep a `List<Pair<Long, LocalItem.LocalItemType>> deletedItems` and when the user deletes an item, add its id+type here
- when doing any change (except deletion), do not update any variable except `debounceSaver.isModified`
- let the debounce saver or `onPause` call `saveImmediate`
- in `saveImmediate`:
  - delete all items found in `deletedItems` and clear `deletedItems`
  - cycle through `items` and if there is an item with a `displayIndex` different than its position in the list, add it to the items to update
  - update the `displayIndex` of items in the list